### PR TITLE
Introduce new type for featurizing Atoms objects

### PR DESCRIPTION
**Pull request by thazhemadam**
Presently, the `featurization` in `AtomGraph` is present for effective serialization support. However, this is not without its issues (see #85).
However, instead of adding a new featurization type, I propose adding a new type that encapsulates the data that would be serialized otherwise, and use this for serialization purposes (and more).
This new type would essentially contain an `AbstractAtoms` object, `AbstractFeaturization` object, and an `encoded_features` field. The features stored in `encoded_features` can always be decoded using the `AbstractFeaturization` associated.

Some of the pros that this approach offers -
* Atoms object becomes more _atomic_
* decreased coupling between  Atoms and Featurization objects
* completely independent, and customisable  serialization (could be specialized for different types if required too, but I think it should be fairly generic)
* no need to add a new pseudo featurization type that does nothing
* no need to additionally set the field to nothing once a convolution op gets performed on an Atoms object
* relaxes the constraint on new types of `AbstractAtom`, as they won't need to have the `featurization` field
* Present serialization logic as present in the constructor for `AtomGraph`  can be simplified. This would also apply to other types of `AbstractAtoms` types (as they're developed).

### CODE REVIEW

File: src/atoms/atomgraph.jl
Comment (rkurchin) [inline src/atoms/atomgraph.jl:82]: I was just playing with something earlier and realized it would be nice if this defaulted to the filename (without extension) when building from file, rather than a blank string. Could you add that?

File: src/atoms/atomgraph.jl
Comment (rkurchin) [inline src/atoms/atomgraph.jl:101]: wait then shouldn't this be `if !isempty(id)`?

File: src/featurizations/graphnodefeaturization.jl
Comment (rkurchin) [inline src/featurizations/graphnodefeaturization.jl:115]: I think this should actually be an alias for the constructor of the new container object. Because otherwise it's not featurizing (i.e. attaching encoded features to a structure), it's just encoding...

File: src/serialize.jl
Comment (rkurchin) [inline src/serialize.jl:4]: Personally, I prefer something like `FeaturizedAtoms` as the name. It's a bit more general and descriptive of what it actually is rather than just one intended purpose of it, and there's nothing about the `Atoms` objects that makes them _not_ serializable...

Comment (rkurchin) [reply src/serialize.jl:4]: Also, it stores multiple encoded features, not just one...

File: src/serialize.jl
Comment (rkurchin) [inline src/serialize.jl:6]: Can we change this to something more general too, like

> Container object for an `Atoms` object, a featurization, and the resulting encoded features from applying the featurization to the atoms.

File: test/featurizations/GraphNodeFeaturization_tests.jl
Comment (rkurchin) [inline test/featurizations/GraphNodeFeaturization_tests.jl:24]: can we make sure it works if the second argument isn't a vector, i.e. I want to featurize multiple structures with the same featurization?

File: src/serialize.jl
Comment (rkurchin) [inline src/serialize.jl:23]: perhaps worth making the new container (currently SEF) a parametric type so that we can dispatch on the types of `Atoms` and featurizations in case it sometimes isn't?

File: test/featurizations/GraphNodeFeaturization_tests.jl
Comment (rkurchin) [inline test/featurizations/GraphNodeFeaturization_tests.jl:31]: I'm not sure I understand why what was there before wouldn't work?

**Review by rkurchin — CHANGES_REQUESTED** [review]
Looking mostly good so far. A couple of comments/questions/suggestions sprinkled throughout.

File: src/atoms/atomgraph.jl
Comment (thazhemadam) [reply src/atoms/atomgraph.jl:101]: Oh yes, it should. Sorry about that. 😅

File: src/serialize.jl
Comment (thazhemadam) [reply src/serialize.jl:4]: In that case, `FeaturizedAtoms` it is! I wasn't really attached to `SerializableEncodedFeature` anyways #ntih

File: src/serialize.jl
Comment (thazhemadam) [reply src/serialize.jl:6]: Yes, that makes more sense to me.

File: src/serialize.jl
Comment (thazhemadam) [reply src/serialize.jl:23]: > perhaps worth making the new container (currently SEF) a parametric type so that we can dispatch on the types of Atoms and featurizations in case it sometimes isn't?

I'm not sure I follow. In case it sometimes isn't what?

Comment (rkurchin) [reply src/serialize.jl:23]: haha sorry, I meant having a parametric type like `SerializableEncodedFeature{AtomsType, FeaturizationType}` and then we can dispatch the decode function on those types if needed in case that loop ends up being not universally applicable.

File: test/featurizations/GraphNodeFeaturization_tests.jl
Comment (thazhemadam) [reply test/featurizations/GraphNodeFeaturization_tests.jl:31]: Only `Block` seems to be tested. Wouldn't it be more robust if all features defined in whatever `fnames` is passed to `GNF` could somehow be tests?

FWIW, we might need to think of a cleaner way to structure these tests too, because these tests for `GraphNodeFeaturization` are very similar to those the new container type (`FeaturizedAtoms`) would require.

File: src/atoms/atomgraph.jl
Comment (rkurchin) [reply src/atoms/atomgraph.jl:101]: 🥁 more tests more tests

File: src/featurizations/graphnodefeaturization.jl
Comment (thazhemadam) [reply src/featurizations/graphnodefeaturization.jl:115]: Isn't it doing that in the inner constructor of `SerializableEncodedFeature`, while the container object is being created?
Or am I misunderstanding your point here?

File: src/atoms/atomgraph.jl
Comment (thazhemadam) [reply src/atoms/atomgraph.jl]: Resolved in 7257df3.

File: src/atoms/atomgraph.jl
Comment (thazhemadam) [reply src/atoms/atomgraph.jl]: Resolved in 7257df3. (for now)
We won't need this after all, as the `id` is set to `input_file_path` now.
If someone wants to intentionally have an empty string in there, I say let them. 😆

File: test/featurizations/GraphNodeFeaturization_tests.jl
Comment (thazhemadam) [reply test/featurizations/GraphNodeFeaturization_tests.jl:27]: Resolved in 9834174.

File: src/serialize.jl
Comment (thazhemadam) [reply src/serialize.jl]: Resolved in 18893cf9dad7d546f3eff93a005abe212957440d.

File: src/serialize.jl
Comment (thazhemadam) [reply src/serialize.jl]: Resolved in 18893cf9dad7d546f3eff93a005abe212957440d.

File: src/serialize.jl
Comment (thazhemadam) [reply src/serialize.jl]: Resolved in fa8595b.

File: src/atoms/atomgraph.jl
Comment (rkurchin) [inline src/atoms/atomgraph.jl:82]: can we strip the extension from this? `splitext` should do it

File: src/featurizations/graphnodefeaturization.jl
Comment (rkurchin) [reply src/featurizations/graphnodefeaturization.jl:115]: Yes, this is purely a naming thing. If we want a function that does what this one does, I think it should be called something like `encode`, and a function called `featurize` should actually do the featurizing (i.e. attaching of encoded features to Atoms object), which would mean that it should essentially be an alias of the `FeaturizedAtoms` constructor. While redundant, I think it probably makes sense to have such a function because it's a pretty intuitive syntax that might be easier for a user to remember.

File: test/featurizations/GraphNodeFeaturization_tests.jl
Comment (rkurchin) [reply test/featurizations/GraphNodeFeaturization_tests.jl:34]: yeah it would probably make sense to test more features, and also agreed about restructuring the tests. I'll leave it up to you if you want to do that before merging or just file an issue to remember to do it later

File: test/featurizations/GraphNodeFeaturization_tests.jl
Comment (rkurchin) [inline test/featurizations/GraphNodeFeaturization_tests.jl:24]: I think there's a way to overwrite the way broadcasts/iterators/whatever work to be able to do it even if it's not in brackets...

**Review by rkurchin — CHANGES_REQUESTED** [review]
just a couple more tiny things

File: src/featurizations/graphnodefeaturization.jl
Comment (thazhemadam) [reply src/featurizations/graphnodefeaturization.jl:115]: Ah, got it. Slightly satisfying that we can now have both `decode` AND `encode` haha

File: test/featurizations/GraphNodeFeaturization_tests.jl
Comment (thazhemadam) [reply test/featurizations/GraphNodeFeaturization_tests.jl:34]: Add it to the Project board, and we'll get to it? I think it might be a better idea to work on those changes in a different PR.

File: test/featurizations/GraphNodeFeaturization_tests.jl
Comment (thazhemadam) [reply test/featurizations/GraphNodeFeaturization_tests.jl:24]: There is?
I think `featurize.([triangle_C_1, triangle_C_2], fzn1)` throws an error.

Comment (rkurchin) [reply test/featurizations/GraphNodeFeaturization_tests.jl:24]: Yeah it definitely will right now. Damn, I think I asked about this exact thing on the Slack awhile ago and now I can't remember what the clean way to do it was...okay, well this shouldn't be a showstopper for now though I definitely do want to figure it out so another one for the project board haha

File: src/atoms/atomgraph.jl
Comment (thazhemadam) [reply src/atoms/atomgraph.jl]: Resolved in b049e0f.

File: src/featurizations/graphnodefeaturization.jl
Comment (thazhemadam) [reply src/featurizations/graphnodefeaturization.jl]: Resolved in 6a61bf5 and 676ee63.

File: src/atoms/atomgraph.jl
Comment (rkurchin) [inline src/atoms/atomgraph.jl:82]: sorry can we actually do this as `splitext(splitpath(input_file_path)[end])[begin]`? i.e. just the file name, not the beginning of the path or the extension

File: src/featurizedatoms.jl
Comment (rkurchin) [inline src/featurizedatoms.jl:37]: can we add docstrings for this and `decode`?

**Review by rkurchin — COMMENTED** [review]
Two very tiny largely aesthetic things and then I'm good to go with this!

File: src/atoms/atomgraph.jl
Comment (thazhemadam) [reply src/atoms/atomgraph.jl:82]: Is it possible that you'd want to build `AtomGraph`s from two files that have the same file name, but reside in different directories in a batch? If it is, then this might make it difficult to distinguish between the two `AtomGraph`'s (if required).
I wasn't entirely sure about that, so I thought I'd leave it as-is 😅

File: src/featurizedatoms.jl
Comment (thazhemadam) [reply src/featurizedatoms.jl:49]: Resolved in 99dd901.

File: src/atoms/atomgraph.jl
Comment (rkurchin) [reply src/atoms/atomgraph.jl:82]: I mean, this is just for the default `id` value. Obviously the user can specify themselves if they want something else. I suppose a future option we could consider would be something like specifying a prefix and otherwise using the default, but I don't think that's necessary right now...anyway, I'm fine leaving this as-is too, it's not a big deal, just wanted some sensible default in there that was more useful than the empty string, which I put in ages ago and forgot to change :P

**Review by rkurchin — APPROVED** [review]

File: src/featurizedatoms.jl
Comment (DhairyaLGandhi) [inline src/featurizedatoms.jl:15]: Why is this constructor internal?

Comment (rkurchin) [reply src/featurizedatoms.jl:15]: Because we want to enforce that the `encoded_features` are actually produced by the `featurization` encoding on the `atoms`. If it were external (assuming I'm understanding correctly), then there would still be a constructor that allows those three to all be fed in independently...

**Pull request merged**
